ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java

```diff
@@ -17,51 +17,52 @@
 import io.confluent.ksql.function.AggregateFunctionFactory;
 import io.confluent.ksql.function.FunctionRegistry;
 import io.confluent.ksql.function.KsqlAggregateFunction;
-import io.confluent.ksql.function.KsqlFunction;
 import io.confluent.ksql.function.UdfFactory;
 import io.confluent.ksql.util.KsqlException;
 import io.confluent.ksql.util.KsqlReferentialIntegrityException;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.concurrent.ThreadSafe;
 import org.apache.kafka.connect.data.Schema;
 
+@ThreadSafe
```

Comment:
+1

Reply:
I'm against the use of `synchronize` on method signatures. It leaks internal implementation details of the class into the API, and can cause issues if other code choosing to `synchronize(metaStore)`.  

This is why I've added `Object` lock objects in other places. However, for this instance I disagree we should add a lock around all mutable state. It's accessed independently and adding course grained concurrency control won't add anything.

Reply:
I'm ambivalent about whether we use the synchronized on the method signatures or just use a synchronized block, but we shouldn't ever allow for mixing state and synchronization by locking on something like `metastore` - that's criminal! 

Interesting point, though, about it leaking into the API, didn't really think about that before. Would that still be a concern if the method is only accessed via an interface that doesn't specify `synchronized`?

Reply:
> It leaks internal implementation details of the class into the API

I don't buy this - we're actually losing some of the benefits of encapsulation: readability and being less bug-prone. It's also arguable that its leaking into the API. The interface defines the API, so it would seem fine for the implementation classes to use synchronized.

> can cause issues if other code choosing to `synchronize(metaStore)`

Can you elaborate on this point? synchronized is reentrant so it should be a problem to use metastore from inside `synchronize(metaStore)` somehow. Are you worried about getting deadlocked somehow?

Reply:
> Interesting point, though, about it leaking into the API, didn't really think about that before. Would that still be a concern if the method is only accessed via an interface that doesn't specify `synchronized`?

As long as the type is only ever accessed via the interface, yeah, then that's better - but people can still do things like `synchonrize(metaStore)`.

Reply:
> Are you worried about getting deadlocked somehow?

Yes, I've seen weird A-B dead locking issues before, hence I always sync on an private object.

Reply:
> I don't buy this - we're actually losing some of the benefits of encapsulation:

Not sure how this is related to encapsulation...

Reply:
> It's also debatable that its leaking into the API - the interface defines the API

In this instance, yes.  I was talking more in general terms.

My other concern stands - i,e. it's a safer pattern to use private locks.